### ISSUE TITLE
Provide a catalog to `sql::pure`?

### ISSUE THREAD

**Issue by sploiselle** [MaterializeInc/materialize#3402]
The `coord`'s catalog could be useful for purifying SQL statements, e.g. I ran into a need for this implementing `CREATE DEFAULT INDEX ON foo`, where ideally, I would have been able to run the `CreateIndex` statement through `pure` to determine the index's structure based on its catalog. However, the `coord`'s catalog is not plug-and-play with `sql::catalog`, so there isn't direct interoperability.  Instead, I relied on a version of the catalog attached to `sql::plan::statement::StatementContext`, which meant we're doing the purification during planning.

**Comment by benesch** [comment]
I apologize for misleading you when you first asked me about this, but this is actually a "working as designed", and I'd completely forgotten. Probably a docs bug somewhere. Anyway, purification is an asynchronous step that _intentionally_ doesn't hold a lock on the catalog, so to speak. This is implicit in the fact that coordinator is single threaded, but essentially planning a SQL statement proceeds like so:

1. coord acquires catalog lock
2. planner plan SQL statement, possibly reading bits of information from the catalog
3. coord applies plan to catalog, if any changes are required
4. coord releases catalog lock

Now, there isn't actually an explicit catalog lock, but the fact that the coordinator processes queries one at a time means that you never have to worry about catalog skew between when the plan is produced and when the plan is applied to the catalog.

In the bad old days, if your query depended on e.g. the Confluent Schema Registry, the steps would look like this:

1. coord acquires catalog lock
2. planner plan SQL statement, possibly reading bits of information from the catalog _and_ the CSR
    * Possibly we'd block forever here waiting for the Schema Registry to return. Even in the best case scenario, we'd prevent any queries from being planned for at least a few milliseconds!
3. coord applies plan to catalog, if any changes are required
4. coord releases catalog lock

Purification exists to hoist that expensive roundtrip with the CSR outside of the critical section, so that we're never preventing any other queries from planning while chatting with the CSR.

tl;dr the whole point of purification is to not have a catalog available.

All that said, possibly it could make sense to note that planning itself happens in two phases: something I'll call normalization, followed by... idk, the rest of planning. Normalization is the process of producing a `create_sql` command; and then there's everything that happens after.

I think the key insight here is that planning is not idempotent. Purification removes _some_ of the nondeterminism, but in the case of `CREATE DEFAULT INDEX`, not all. I think "purification" is probably an OK word because `CREATE DEFAULT INDEX` is still "pure" in the sense that it doesn't depend on any external state; it just depends on the internal state of the catalog. So then normalization removes the last bit of idempotency.

So the pipeline is roughly:

  * Parsing, which transforms a SQL string into an AST
  * Purification, which transforms an AST -> AST that does not depend on any *external* state
  * Normalization, which transforms an AST -> AST that is *idempotent*
  * "Planning" (for want of a better word), which transforms a purified, idempotent AST -> correlated IR 
  * Decorrelation, which transforms correlated IR -> uncorrelated IR
  * Transform (oy, naming), which transforms uncorrelated IR -> uncorrelated IR to optimize it
  * Render, which builds a dataflow graph from an optimized uncorrelated IR

Anyway, definitely up for getting this written down closer to the code in question, though I think the fact that we don't have a separate `normalization` function as part of the SQL API is probably correct; unlike purification which needs to tie into the control flow of the coordinator, normalization is something that should always happen implicitly in the planner.

**Comment by sploiselle** [comment]
@benesch 🙇 ty--updated this ticket to get these insights documented in the code.
